File: Libs/source/network/rudp/segment/ExtendAckSegment.cpp
#include "network/rudp/segment/ExtendAckSegment.h"

/*
*  EACK Segment
*
*   0 1 2 3 4 5 6 7 8            15
*  +-+-+-+-+-+-+-+-+---------------+
*  |0|1|1|0|0|0|0|0|     N + 6     |
*  +-+-+-+-+-+-+-+-+---------------+
*  | Sequence #    |   Ack Number  |
*  +---------------+---------------+
*  |1st out of seq |2nd out of seq |
*  |  ack number   |   ack number  |
*  +---------------+---------------+
*  |  . . .        |Nth out of seq |
*  |               |   ack number  |
*  +---------------+---------------+
*  |            Checksum           |
*  +---------------+---------------+
*
*/
namespace izanagi {
namespace network {
    ExtendAckSegment::ExtendAckSegment(
        IMemoryAllocator* allocator,
        IZ_INT sequenceNumber,
        IZ_INT acknowledgedNumber,
        IZ_INT* ackNumbers,
        IZ_UINT ackNumbersNum)
        : Segment(Type::EAK, sequenceNumber, RUDP_HEADER_LEN + ackNumbersNum)
    {
        m_AckNumbers.init(allocator, ackNumbersNum);

        for (IZ_UINT i = 0; i < ackNumbersNum; i++)
        {
            m_AckNumbers.at(i) = ackNumbers[i];
        }
    }

    void ExtendAckSegment::WriteBytes(IOutputStream* wr)
    {
        Segment::WriteBytes(wr);

        m_AckNumbers.traverse(
            [&](IZ_INT& n) {
            wr->Write((IZ_UINT8)n);
        });

        // TODO
        // padding...

        // For CheckSum
        wr->Write(0);
        wr->Write(0);
    }

    void ExtendAckSegment::ReadBytes(
        IMemoryAllocator* allocator,
        IZ_UINT8* bytes, IZ_UINT offset, IZ_UINT length)
    {
        Segment::ReadBytes(allocator, bytes, offset, length);

        IZ_ASSERT(CheckFlag(Type::EAK));

        auto dataLength = length - RUDP_HEADER_LEN;
        m_AckNumbers.init(allocator, dataLength);

        for (IZ_UINT i = 0; i < dataLength; i++) {
            m_AckNumbers.at(i) = bytes[offset + RUDP_HEADER_LEN + i];
        }
    }
}   // namespace network
}   // namespace izanagi

File: Trials/Network/RUDP/segment/ExtendAckSegment.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

/*
 *  EACK Segment
 *
 *   0 1 2 3 4 5 6 7 8            15
 *  +-+-+-+-+-+-+-+-+---------------+
 *  |0|1|1|0|0|0|0|0|     N + 6     |
 *  +-+-+-+-+-+-+-+-+---------------+
 *  | Sequence #    |   Ack Number  |
 *  +---------------+---------------+
 *  |1st out of seq |2nd out of seq |
 *  |  ack number   |   ack number  |
 *  +---------------+---------------+
 *  |  . . .        |Nth out of seq |
 *  |               |   ack number  |
 *  +---------------+---------------+
 *  |            Checksum           |
 *  +---------------+---------------+
 *
 */
namespace RUDP
{
    class ExtendAckSegment : Segment
    {
        public override string GetSegmentType()
        {
            return "EAK";
        }

        public byte[] AckNumbers
        {
            get;
            private set;
        }

        public ExtendAckSegment()
        {
        }

        public ExtendAckSegment(
            int sequenceNumber,
            int acknowledgedNumber,
            int[] ackNumbers)
            : base(Type.EAK, sequenceNumber, RUDP_HEADER_LEN + ackNumbers.Length)
        {
            this.AckNumbers = new byte[ackNumbers.Length];
            
            for (int i = 0; i < ackNumbers.Length; i++)
            {
                this.AckNumbers[i] = (byte)ackNumbers[i];
            }
        }

        public override void WriteBytes(Stream wr)
        {
            base.WriteBytes(wr);

            wr.Write(this.AckNumbers, 0, this.AckNumbers.Length);

            // For CheckSum
            wr.WriteByte(0);
            wr.WriteByte(0);
        }

        public override void ReadBytes(byte[] bytes, int offset, int length)
        {
            base.ReadBytes(bytes, offset, length);

            if (!CheckFlag(Type.EAK))
            {
                throw new ArgumentOutOfRangeException();
            }

            var dataLength = bytes.Length - RUDP_HEADER_LEN;
            this.AckNumbers = new byte[dataLength];

            System.Array.Copy(
                bytes,
                offset + RUDP_HEADER_LEN,
                this.AckNumbers,
                0,
                this.AckNumbers.Length);
        }
    }
}
